### DISCUSSION TITLE
Multiple copies of the same device, only 1 device in HA?

### DISCUSSION THREAD

**Discussion by rrozema** [dawidchyrzynski/arduino-home-assistant#149] [Q&A]
I've made a small test device 'stoplicht' (which is Dutch for 'traffic light', so you'll have a good idea now what exactly this device does). I've got the code in place to give each device a unique id using ESP.getChipID(). I've uploaded the exact same code into multiple WEMOS D1 mini's and ran them. If I look in MQTT Explorer I do see each individual device listed using their unique ids. In Home Assistant however only the device I started first does get entities created for it. I had expected each copy to get their own entities created in HA. In the log I do see an error message for each entity of the next copies:

> 2023-03-17 23:34:06.129 ERROR (MainThread) [homeassistant.components.switch] Platform mqtt does not generate unique IDs. ID red already exists - ignoring switch.red

I'm only listing the error for the entity 'red' here, the other entities have the same error.

I've found that a unique config message does get created in MQTT for each entity for each (copied) device. In the config listed in f.e. homeassistant/switch/dbfa2b00/red/config however the message sent has the same value for "uniq_id": "red" for each device. f.e. Here is the config listed for one device's "red" entity:
`{
  "name": "red",
  "uniq_id": "red",
  "ic": "mdi:led-outline",
  "dev": {
    "ids": "dbfa2b00",
    "name": "StopLicht_2BFADB",
    "sw": "1.0.0",
    "mf": "Richard Rozema",
    "mdl": "stoplicht"
  },
  "avty_t": "stoplicht/dbfa2b00/avty_t",
  "stat_t": "stoplicht/dbfa2b00/red/stat_t",
  "cmd_t": "stoplicht/dbfa2b00/red/cmd_t"
}`
And here is the config for another device's red entity:
`{
  "name": "red",
  "uniq_id": "red",
  "ic": "mdi:led-outline",
  "dev": {
    "ids": "f8abe000",
    "name": "StopLicht_E0ABF8",
    "sw": "1.0.0",
    "mf": "Richard Rozema",
    "mdl": "stoplicht"
  },
  "avty_t": "stoplicht/f8abe000/avty_t",
  "stat_t": "stoplicht/f8abe000/red/stat_t",
  "cmd_t": "stoplicht/f8abe000/red/cmd_t"
}`

I expected the uniq_id passed into HA for each entity to include the unique id I specified for the device, so that each entity would have a unique id in HA too. Do I really need to explicitly include the device's unique_id in each entity's unique id or did I do something wrong?

**Comment by justind000** [comment]
Have a look at this [issue](https://github.com/dawidchyrzynski/arduino-home-assistant/issues/124). It shows how to do what you're trying to do.

**Reply by rrozema** [reply]
No, I already call the HADevice.setUniqueID() method. That however only makes the topics in "dev": unique, not the "uniq_id":, which is specific for the entity. If you look at the both "config" entries I posted, you'll see that the first copy has "ids": "dbfa2b00", and all the releated fields ("name", "name", "avty_t", "stat_t" and "cmd_t") are all correctly set to make the config unique in MQTT. However the "uniq_id" still shows the same value of "red" for both copies (and "green", "yellow" and "stoplicht" for the other entities). HA apparently needs the value in "uniq_id" to be unique in the HA-instance too. So it seems HA expects something like this (with the device's "ids" value prepended to the entity's "uniq_id" value):
`{ "name": "red", "uniq_id": "dbfa2b00-red", "ic": "mdi:led-outline", "dev": { "ids": "dbfa2b00", "name": "StopLicht_2BFADB", "sw": "1.0.0", "mf": "Richard Rozema", "mdl": "stoplicht" }, "avty_t": "stoplicht/dbfa2b00/avty_t", "stat_t": "stoplicht/dbfa2b00/red/stat_t", "cmd_t": "stoplicht/dbfa2b00/red/cmd_t" }`
So, I think that that issue is indeed part of the solution, but it is not complete yet.

**Reply by justind000** [reply]
Have a look at the second to last entry in the referenced issue. It creates a new sensor object with a concatd name. Pretty sure that's the trick.

You also need globally unique sensor names, regardless of device name they are attached to.

**Reply by rrozema** [reply]
Actually, to be exact: I do NOT call the HADevice.setUniqueID() method. I call the HASwitch(const char* uniqueId) constructor, which calls the HADevice(const char* uniqueId) constructor. But the exact same thing happens when you call the setUniqueID() method.  (I'm not an experienced c++ developer, but) looking at the HADevice.c source file, it seems to me that indeed at line 25 the value passed in for uniqueId is passed on directly into the _serializer's HADeviceIndentifiersProperty using this code  `_serializer->set(AHATOFSTR(HADeviceIdentifiersProperty), _uniqueId);`. I was expecting the device's unique Id to be pre-pended either here or at some alternate location just before the serializer gets output into whatever stream it needs to be serialized into. I can not find such pre-pending however in either place.

**Reply by rrozema** [reply]
Yes, you are correct. That is the _work around_. But the cause is in the library not setting a unique id for the entities.

**Reply by rrozema** [reply]
So to conclude, each entity defined in my device needs to be given an id:
- that is unique within the device,
- that is unique within the home assistant instance,
- that remains unchanged over each invocation of my device,
- and that is stored in the device's memory statically for the run-time of my device.
Given that we already need to specify an id for each instance of the device, which needs to be unique within the home assistant instance (suggested to use the device's mac address). Plus that we further need to provide an id for each entity unique within the device. Why does the library not combine these two unique ids into one 'globally unique' id for the entity in Home assistant? Even worse, not only does the library not help us by generating such a 'globally unique' entity id from these two pieces of information that are already available to it, but it makes it even harder on us by imposing an additional restriction that the memory we use to pass in this id that we have to hand-build, must remain static in memory of the device for the entire duration of the run time of the device.

I really like most of this library, but this little bit of the entity id makes no sense at all to me ...

**Comment by jwilleke** [comment]
I too found this confusing and had to go back and read the [documentation](https://dawidchyrzynski.github.io/arduino-home-assistant/documents/library/device-configuration.html) which says 
_HADevice represents the physical device where the library is installed. Logically it’s a group of types like sensors, switches, lights and so on. In the Home Assistant, it’s listed with properties that may be configured using the library’s API.
...
The ID of a device needs to be unique in a scope of a Home Assistant instance. The safest solution is to use the MAC address of an Ethernet or Wi-Fi chip but you can also implement your own solution.
... then under [Device Types](https://dawidchyrzynski.github.io/arduino-home-assistant/documents/library/device-types.html)
Device type represents a single entity in the Home Assistant panel. It can be a sensor, lock, camera or anything that’s listed in the table below_.
_
The HADevice and UniqueID is the device (in home assistant) which which hosts IP Address needs to be declared before setup().
```HADevice device(myId, sizeof(myId));``` (I used the mac address)

Then your various [Device Types](https://dawidchyrzynski.github.io/arduino-home-assistant/documents/library/device-types.html) are "Entities" (in home assistant) and can be defined  similar to like:
```HASensorNumber uptimeSensor("ardUptime"); //<-- UniqueID of Sensor Type```


Topic: homeassistant/sensor/f412faa081c0/ardUptime/config
Payload
{
  "name": "XXXUptime",
  "unique_id": "ardUptime",
  "icon": "mdi:home",
  "unit_of_measurement": "s",
  "device": {
    "identifiers": "f412faa081c0",
    "sw_version": "1.0.0",
    "manufacturer": "arduino",
    "model": "uno_r4_wifi"
  },
  "availability_topic": "aha/f412faa081c0/avty_t",
  "state_topic": "aha/f412faa081c0/ardUptime/stat_t",
  "platform": "mqtt"
}

Let me know how I can help.
